### DISCUSSION TITLE
How to specify the network interface used in our proxy?

### DISCUSSION THREAD

**Discussion by LoickZoty** [abhinavsingh/proxy.py#1416] [Q&A] [answered]
I have two distinct local networks, 192.168.1.0/24 and 10.1.100.0/24.

Let's say I run a proxy like this:
```python
if __name__ == '__main__':
  proxy.main(
    hostname="192.168.1.150",
    port=8899
  )
```
This listens on my first network, 192.168.1.150, which is fine for now.

But how can I make the proxy redirect traffic to the second network, 10.1.100.10? Is it possible to specify the network interface in some way?

I made a small diagram:
```
| User uses a proxy      |       ------->      | Proxy input on 192.168.1.150:8899 | 
                                                                                               |
                                                                                               |
                                                                                               |
                                                                                               |
| Response to the user  |       <-------      | Proxy redirects traffic to 10.1.100.10 | 
```

I've tried looking into the source code, and I think my answers lie in 'proxy/core/connection'. However, I'm having trouble figuring out how to use them with override or via a plugin.

Does anyone have any ideas on how to achieve this ?

**Comment by abhinavsingh** [comment]
@LoickZoty What do you mean by redirect traffic to 2nd network?   Your code listening on `192.168.1.150` can simply hit services under `10.1.100.10` and response back as usual.  Also, I think you are referring to overriding source IP in the internals.   Don't think this is what you need.  IIUC, you simply want a request on `192.168.*` to be served from services available under `10.1.*`, right?

**Reply by LoickZoty** [reply]
That's exactly it!

I've made some progress on my issue and it seems like I can achieve my goal by using this (via a plugin):
```python
    def resolve_dns(self, host: str, port: int) -> Tuple[Optional[str], Optional['HostPort']]:
        return None, ("192.168.1.150", 56925) #
        # return None, None
```

This will allow defining the source_address in 'proxy/common/utils.py':
```python
def new_socket_connection(
        addr: HostPort,
        timeout: float = DEFAULT_TIMEOUT,
        source_address: Optional[HostPort] = None,
) -> socket.socket:
    conn = None
    ...
```

When I send a request through the proxy, I get a "(502) Bad Gateway" error. This error is puzzling because it is the same IP address as if I returned None, None in resolve_dns(...) (it's my default network).

At least, maybe I'm on the wrong way haha.

**Reply by LoickZoty** [reply, marked as answer]
The response:

You needed to set the port to '0' so that the machine could choose a random port. Otherwise, multiple sockets would be created with the specified port, causing an error!

```
def resolve_dns(self, host: str, port: int) -> Tuple[Optional[str], Optional['HostPort']]:
    return None, ("192.168.1.150", 0)
```
